tools/chplcheck/src/rule_types.py

```diff
@@ -113,12 +126,150 @@ def fixits(self, context: chapel.Context, name: str) -> typing.List[Fixit]:
     typing.Union[chapel.AstNode, AdvancedRuleResult]
 ]
 """Internal type for advanced rule results"""
-AdvancedRule = typing.Callable[
+
+
+AdvancedRuleCheck = typing.Callable[
     [chapel.Context, chapel.AstNode], _AdvancedRuleResult
 ]
 """Function type for advanced rules"""
 
+
 RuleResult = typing.Union[_BasicRuleResult, _AdvancedRuleResult]
 """Union type for all rule results"""
-Rule = typing.Union[BasicRule, AdvancedRule]
-"""Union type for all rules"""
+
+CheckResult = typing.Tuple[chapel.AstNode, str, typing.List[Fixit]]
+
+
+VarResultType = typing.TypeVar("VarResultType")
+
+FixitHook = typing.Callable[
+    [chapel.Context, VarResultType],
+    typing.Optional[typing.Union[Fixit, typing.List[Fixit]]],
+]
+"""
+Function type for fixits; (context, data) -> None or Fixit or List[Fixit]
+"""
+
+
+class Rule(typing.Generic[VarResultType]):
+    # can't specify type of driver due to circular import
+    def __init__(self, driver, name: str) -> None:
```

Comment:
I'd also consider moving `check_func` into the superclass

Reply:
They have different signatures, and I'd need to add a second type parameter since they're an `_AdvancedRuleResult` (union of possible return values) and not an `AdvancedRuleResult` (just the class). I think it's messier to move it up.

Reply:
Works for me